datalad/support/sshconnector.py

```diff
@@ -207,6 +214,43 @@ def runner(self):
             self._runner = WitlessRunner()
         return self._runner
 
+    @property
+    def ssh_version(self):
+        """
+        Determine name and version of the ssh-client.
+
+        Returns
+        -------
+        Tuple[str, int, int, int]
+          - first element of the result is the name of the client or '?' if the
+            client is unknown
+          - second element of the result is the version number or 0 if the
+            client is unknown
+          - third element of the result is the sub-version number or 0 if the
+            client is unknown
+          - fourth element of the result is the patch number or 0 if the
+            client is unknown
+        """
+        if self._ssh_version is None:
+            result = self.runner.run(
+                [self.ssh_executable, "-V"],
+                protocol=StdErrCapture)
```

Comment:
`self.ssh_executable` is local, isn't it?  I would probably not rely on that it would match location on remote, could be a different OS even. Just `ssh`?

or may be we should avoid doing that at all and just sense it from `ssh -v` output?

```
$> ssh -v localhost echo 2>&1 | grep 'remote software'
debug1: Remote protocol version 2.0, remote software version OpenSSH_8.4p1 Debian-5
```
?

Reply:
W.r.t. remote end ssh-version: I tested the following configurations

local ssh8 -> remote ssh8: quoting required
local ssh8 -> remote ssh9: quoting required
local ssh9 -> remote ssh8: no quoting required
local ssh9 -> remote ssh9: no quoting required

It seems to be sufficient to detect the local OpenSSH version in order to decide for or against quoting.

Reply:
I wonder why in https://github.com/datalad/datalad/issues/6655 I have made a conclusion "seems to be largely server (not client) version dependent! " ... I guess because when I scped `uco py` (no `\`) to local which had 9.0 I got `/tmp/uc\ opy`, whenever when I scp'ed to smaug which had older 8.4, I got it Ok. (I would not be surprised if some incompatibility is still lurking in scp itself).

can you confirm that with 9->9 you do get an explicit `\` in filename if you try repeating experiments I have done?

Reply:
or may be it is just `ls` also did some "trickery" in local vs smaug in my cases... uff

Reply:
The experiment with 9->9 (remote is running on localhost:11001):

```shell
> /usr/local/bin/ssh -V
OpenSSH_9.0p1, OpenSSL 1.1.1d  10 Sep 2019
> /usr/local/bin/ssh -p 11001 -v localhost echo 2>&1 | grep 'remote software'
debug1: Remote protocol version 2.0, remote software version OpenSSH_9.0
> rm -rf /tmp/uc*; /usr/local/bin/scp -P 11001 ab 'localhost:/tmp/uc\ opy'; /usr/local/bin/ssh -p 11001 localhost 'ls -ld /tmp/uc*'
-rw-r--r-- 1 cristian users 24 Jul  8 20:18 /tmp/uc\ opy
```

**Confirmed!**



The experiment with 9->8:

```shell
> /usr/local/bin/ssh -V
OpenSSH_9.0p1, OpenSSL 1.1.1d  10 Sep 2019
> /usr/local/bin/ssh -v localhost echo 2>&1 | grep 'remote software'
debug1: Remote protocol version 2.0, remote software version OpenSSH_8.4
> /usr/local/bin/ssh localhost /usr/bin/ssh -V
OpenSSH_8.4p1, OpenSSL 1.1.1d  10 Sep 2019
> rm -rf /tmp/uc*; /usr/local/bin/scp ab 'localhost:/tmp/uc\ opy'; /usr/local/bin/ssh localhost 'ls -ld /tmp/uc*'
-rw-r--r-- 1 cristian users 24 Jul  8 20:13 /tmp/uc\ opy
```
**Also confirmed!**

Reply:
<details>
<summary>reconfirmed that you are right -- only local ssh version seems to matter</summary> 

```shell
$> rm -rf /tmp/uc*; ssh -V; /usr/bin/scp -v ab 'localhost:/tmp/uc\ opy' 2>&1 | grep software ; /usr/bin/ssh localhost 'ls -ld /tmp/uc*'
OpenSSH_9.0p1 Debian-1+b1, OpenSSL 3.0.3 3 May 2022
debug1: Remote protocol version 2.0, remote software version OpenSSH_9.0p1 Debian-1+b1
-rw------- 1 yoh yoh 0 Jul  8 14:57 /tmp/uc\ opy
1 35979.....................................:Fri 08 Jul 2022 02:57:59 PM EDT:.
lena:/tmp
$> rm -rf /tmp/uc*; ssh -V; /usr/bin/scp -v ab 'localhost:/tmp/uc opy' 2>&1 | grep software ; /usr/bin/ssh localhost 'ls -ld /tmp/uc*' 
OpenSSH_9.0p1 Debian-1+b1, OpenSSL 3.0.3 3 May 2022
debug1: Remote protocol version 2.0, remote software version OpenSSH_9.0p1 Debian-1+b1
-rw------- 1 yoh yoh 0 Jul  8 14:58 /tmp/uc opy
1 35980.....................................:Fri 08 Jul 2022 02:58:09 PM EDT:.
lena:/tmp
$> rm -rf /tmp/uc*; ssh -V; /usr/bin/scp -v ab 'smaug:/tmp/uc opy' 2>&1 | grep software ; /usr/bin/ssh smaug 'ls -ld /tmp/uc*'    
OpenSSH_9.0p1 Debian-1+b1, OpenSSL 3.0.3 3 May 2022
debug1: Remote protocol version 2.0, remote software version OpenSSH_8.4p1 Debian-5
-rw------- 1 yoh yoh     0 Jul  8 14:58 /tmp/uc opy
-rw------- 1 yoh yoh 28253 Apr 26 18:25 /tmp/uc\ opy
1 35981.....................................:Fri 08 Jul 2022 02:58:28 PM EDT:.
lena:/tmp
$> ssh smaug 'rm -rf /tmp/uc*'; ssh -V; /usr/bin/scp -v ab 'smaug:/tmp/uc opy' 2>&1 | grep software ; /usr/bin/ssh smaug 'ls -ld /tmp/uc*'
OpenSSH_9.0p1 Debian-1+b1, OpenSSL 3.0.3 3 May 2022
debug1: Remote protocol version 2.0, remote software version OpenSSH_8.4p1 Debian-5
-rw------- 1 yoh yoh 0 Jul  8 14:58 /tmp/uc opy
1 35982.....................................:Fri 08 Jul 2022 02:58:41 PM EDT:.
lena:/tmp
$> ssh smaug 'rm -rf /tmp/uc*'; ssh -V; /usr/bin/scp -v ab 'smaug:/tmp/uc\ opy' 2>&1 | grep software ; /usr/bin/ssh smaug 'ls -ld /tmp/uc*'
OpenSSH_9.0p1 Debian-1+b1, OpenSSL 3.0.3 3 May 2022
debug1: Remote protocol version 2.0, remote software version OpenSSH_8.4p1 Debian-5
-rw------- 1 yoh yoh 0 Jul  8 14:58 /tmp/uc\ opy
1 35983.....................................:Fri 08 Jul 2022 02:58:52 PM EDT:.
lena:/tmp
$> rm -rf /tmp/uc*; ssh -V; /usr/bin/scp -v ab 'localhost:/tmp/uc opy' 2>&1 | grep software ; /usr/bin/ssh localhost 'ls -ld /tmp/uc*'     
OpenSSH_9.0p1 Debian-1+b1, OpenSSL 3.0.3 3 May 2022
debug1: Remote protocol version 2.0, remote software version OpenSSH_9.0p1 Debian-1+b1
-rw------- 1 yoh yoh 0 Jul  8 14:59 /tmp/uc opy
1 35984.....................................:Fri 08 Jul 2022 02:59:00 PM EDT:.
lena:/tmp
$> rm -rf /tmp/uc*; ssh -V; /usr/bin/scp -v ab 'localhost:/tmp/uc\ opy' 2>&1 | grep software ; /usr/bin/ssh localhost 'ls -ld /tmp/uc*'
OpenSSH_9.0p1 Debian-1+b1, OpenSSL 3.0.3 3 May 2022
debug1: Remote protocol version 2.0, remote software version OpenSSH_9.0p1 Debian-1+b1
-rw------- 1 yoh yoh 0 Jul  8 14:59 /tmp/uc\ opy

```
</details>

so -- can just use `external_versions['cmd:ssh'] < '9'` although not sure how it would behave in case of not found `ssh`